src/shape.cpp

```diff
@@ -494,6 +504,33 @@ bool shape::any_of_dynamic() const
     });
 }
 
+bool shape::valid() const
+{
+    if(not this->dynamic())
+    {
+        return true;
+    }
+    // check min < max and opts within [min, max]
+    bool is_valid = std::none_of(this->dyn_dims().cbegin(), this->dyn_dims().cend(), [&](auto dd) {
```

Comment:
just return `std::none_of`

Reply:
Inside none_of lambda you can reduce to
`return ((dd.min > dd.max) or 
(std::any_of(dd.opts.cbegin(), dd.opts.cend(), [&](auto o) {
               return (o < dd.min or o > dd.max);
           })));`